zkevm-circuits/src/evm_circuit/execution/precompiles/ec_mul.rs

```diff
@@ -0,0 +1,519 @@
+use bus_mapping::precompile::{PrecompileAuxData, PrecompileCalls};
+use eth_types::{Field, ToLittleEndian, ToScalar, Word};
+use gadgets::util::{and, not, or, Expr};
+use halo2_proofs::{circuit::Value, plonk::Error};
+
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        step::ExecutionState,
+        util::{
+            common_gadget::RestoreContextGadget,
+            constraint_builder::{ConstrainBuilderCommon, EVMConstraintBuilder},
+            math_gadget::{IsZeroGadget, ModGadget},
+            CachedRegion, Cell, RandomLinearCombination,
+        },
+    },
+    table::CallContextFieldTag,
+    witness::{Block, Call, ExecStep, Transaction},
+};
+
+// Modulus for scalar field Fr
+const FR_N: [u8; 32] = [
+    0x01, 0x00, 0x00, 0xF0, 0x93, 0xF5, 0xE1, 0x43, 0x91, 0x70, 0xB9, 0x79, 0x48, 0xE8, 0x33, 0x28,
+    0x5D, 0x58, 0x81, 0x81, 0xB6, 0x45, 0x50, 0xB8, 0x29, 0xA0, 0x31, 0xE1, 0x72, 0x4E, 0x64, 0x30,
+];
+
+#[derive(Clone, Debug)]
+pub struct EcMulGadget<F> {
+    point_p_x_rlc: Cell<F>,
+    point_p_y_rlc: Cell<F>,
+    scalar_s_rlc: Cell<F>,
+    scalar_s_raw_rlc: Cell<F>,
+    point_r_x_rlc: Cell<F>,
+    point_r_y_rlc: Cell<F>,
+
+    p_x_is_zero: IsZeroGadget<F>,
+    p_y_is_zero: IsZeroGadget<F>,
+    s_is_zero: IsZeroGadget<F>,
+
+    // Two Words (s_raw, s_modded) that satisfies
+    // k * FR_N + s_modded = s_raw
+    // Used for proving correct modulo by Fr
+    s_mod_pair: (
+        RandomLinearCombination<F, 32>, // raw
+        RandomLinearCombination<F, 32>, // mod by FR_N
+    ),
+    n: RandomLinearCombination<F, 32>, // modulus
+    modword: ModGadget<F>,
```

Comment:
The `ModGadget`, i.e.`a mod n == r` also takes in `n` as a "variable" (typically an EVM input).

In our special case here, we already know what `n` is, specifically, `Fr::MODULUS` (or `FR_N` as it is defined in your code).

So we should refactor this `ModGadget` gadget into another `ModByConstGadget` (not really picky about this nomenclature), which does a modulo operation by a constant (as opposed to a variable).

Also, there are conditions like "is a zero?" or "is n zero?" in the `ModGadget` which are not required by us in `ecMul`. So for now, even though using the `ModGadget` is a sound approach, its not the most efficient approach as it needs more cells than necessary, more constrains than necessary.

If you feel its a bit too much to make all those changes in this PR, I am happy to open a tracking issue to do it as a follow-up PR to this one. Wdyt? cc: @lispc

Reply:
I think it's a great idea to add constant mod gadgets. Specifically, it can be named after each curve ex. `ModFrBN256Gadget`. Modulating 256-bit numbers with a curve's `p` or scalar field order `n` is a common operation so the gadget will be reusable in the future. `n` will be hard-coded and zero conditions eliminated.